Customizable folder stripping features

SolidAlloy opened the pull request:
If prefab is instantiated at runtime, its folders will not be stripped. This PR adds an option to strip folders from prefabs for both Play Mode and builds. It is built on top of #14 and #16 because it needs access to stripping modes introduced in #16.

- The plugin adds a "FolderUser" label to prefabs that have folders inside (and removes the labels automatically if the prefab stops using folders).
- Each time the user enters Play Mode, the plugin goes through all prefabs with the label and strips folders from them.
- Upon exiting Play Mode, the plugin restores changes to the prefabs.
- Almost the same process happens on building the project, except that only prefabs that will be packed to the build are edited.
- Options to disable stripping folders from prefabs in Play Mode and on build are added to Preferences.

I decided to edit all prefabs upon entering Play Mode instead of only ones referenced in the scene because a prefab may be hot-swapped in Play Mode.

This PR should be a big game-changer. I prefer keeping hierarchy small and instantiate objects from prefabs instead of keeping them in the scene. With such an approach, if folders are not stripped from prefabs, the plugin is of little use. I haven't bought any paid hierarchy folder plugins but doubt they have such a feature.

- [x] Checked in Unity 2019-2021
- [x] Added information to README

SolidAlloy committed:
Added StripSettings class to manage StrippingMode settings

SolidAlloy committed:
Added SettingsDrawer class to show the settings in Preferences

SolidAlloy committed:
Made StrippingMode values more clear

SolidAlloy committed:
Made Folder.Flatten() handle different StrippingMode values

SolidAlloy committed:
Added optional folder names capitalization when separator is used

SolidAlloy committed:
Fixed typos in Hierarchy naming

SolidAlloy committed:
Refactored Folder.Flatten(), improved readability of the code

SolidAlloy committed:
Added information about stripping modes to README

SolidAlloy committed:
Bumped package version to 0.3.0

SolidAlloy committed:
Added a shader to replace color of the texture

I haven't found a default shader with such behaviour, so made a very simple custom one. It acts the same way as the previous GetWhiteTexture method: sets RGB values of the set color, but leaves alpha of the original texture.

SolidAlloy committed:
Made editor icons use shaders instead of slow color filters

Changing textures with shaders is much more faster than using custom C# methods to manage pixels. New GetTextureWithMaterial method is 16x faster than GetColorizedTexture.

SolidAlloy committed:
Moved icon-related scripts and assets to Editor/Icons folder

SolidAlloy committed:
Added back HideFlags.DontSave

I overlooked that this flag was set for textures in GetColorizedTexture(). It is necessary for the texture to not be unloaded when Play Mode switches to Edit Mode.

SolidAlloy committed:
Started caching materials and moved texture-changing methods to TextureHelper

This shaved off 10-20 more ms.

SolidAlloy committed:
Merge branch 'master' into merged

SolidAlloy committed:
Started adding labels to prefabs that have the folder component

SolidAlloy committed:
Moved label-adding logic from OnWillCreateAsset to OnPostprocessAllAssets and added label-removing logic

OnPostprocessAllAssets is more convenient in that the assets can already be loaded from AssetDatabase without a need for coroutines, and they support more cases (like when changes are applied to prefab from its override)

SolidAlloy committed:
Started stripping folders from prefabs on Play Mode start, then adding them back.

SolidAlloy committed:
Added folders stripping from prefabs before build

SolidAlloy committed:
Strip folders only from prefabs that will make it into the build

SolidAlloy committed:
Moved settings-related scripts to the Settings folder and renamed a few other folders

Icons -> Icon Handling
Prefabs -> Prefab Handling

This will make it clear the folders contain icon-handling scripts and not icons, etc.

SolidAlloy committed:
Started generating keywords from field names

Since more fields are to be added to the settings, it becomes tedious to update the keywords each time. This approach will generate keywords on fly.

SolidAlloy committed:
Added options to disable stripping folders from prefabs in play mode or build

SolidAlloy committed:
Made PrefabFolderStripper respect the options set in Preferences

SolidAlloy committed:
Added information about stripping folders from prefabs to README

SolidAlloy committed:
Made PrefabFolderStripper compatible with Unity 2019

xsduan commented on Editor/Prefab Handling/PrefabFolderStripper.cs:
```diff
@@ -0,0 +1,146 @@
+
+        private static void RevertChanges()
+        {
+            foreach ((string path, string content) in _changedPrefabs)
```
When leaving Play mode:
```
NullReferenceException: Object reference not set to an instance of an object
UnityHierarchyFolders.Editor.PrefabFolderStripper.RevertChanges () (at E:/game projects/unity/unity-hierarchy-folders/Editor/Prefab Handling/PrefabFolderStripper.cs:118)
UnityHierarchyFolders.Editor.PrefabFolderStripper.HandlePrefabsOnPlayMode (UnityEditor.PlayModeStateChange state) (at E:/game projects/unity/unity-hierarchy-folders/Editor/Prefab Handling/PrefabFolderStripper.cs:50)
UnityEditor.EditorApplication.Internal_PlayModeStateChanged (UnityEditor.PlayModeStateChange state) (at <8d21067e8d9c494db25a2b2485216e63>:0)
```
This makes the operation destructive as well.

(2019.4, 2020.1.4f1, 2021.1.0b7)

SolidAlloy replied:
Ah, I tested it with Enter Play Mode Options enabled, so a domain reload didn't happen on entering Play Mode. I fixed the bug by serializing information about changed prefabs in 0096373e0fceb23a7f9f296b912a52b1cbaf3fa2.

SolidAlloy replied:
0096373e0fceb23a7f9f296b912a52b1cbaf3fa2 hasn't fixed the issue completely. It should be finally fixed in 56d51167ea6f292aa8175ba5c843c9f4162f1502

xsduan commented on Editor/Settings/SettingsDrawer.cs:
```diff
@@ -0,0 +1,111 @@
+            { nameof(StripSettings.PlayMode), "Play Mode Stripping Type" },
+            { nameof(StripSettings.Build), "Build Stripping Type" },
+            { nameof(StripSettings.CapitalizeName), "Capitalize Folder Names" },
+            { nameof(StripSettings.StripFoldersFromPrefabsInPlayMode), "Strip folders from prefabs in Play Mode" },
```
![image](https://user-images.githubusercontent.com/9505910/108789241-70c0b900-753f-11eb-817a-e8ed1c0d9b30.png)

This setting gets reset to whatever the build setting is when you hit play 

(2019.4, 2020.1.4f1, 2021.1.0b7)

SolidAlloy replied:
Fixed in 2f7e3193be17ea4df37f3f883cd21a65f7aa3a8b
Settings Manager requires to write a lot of duplicate code, so mistakes like this are easy to make 😔

SolidAlloy replied:
2f7e3193be17ea4df37f3f883cd21a65f7aa3a8b hasn't fixed the issue completely. It should be finally fixed in 56d51167ea6f292aa8175ba5c843c9f4162f1502

xsduan commented on Editor/Prefab Handling/PrefabFolderStripper.cs:
```diff
@@ -0,0 +1,146 @@
+
+        private static void StripFoldersFromPrefab(string prefabPath, StrippingMode strippingMode)
+        {
+            using (var temp = new EditPrefabContentsScope(prefabPath))
```
![image](https://user-images.githubusercontent.com/9505910/108790131-9484fe80-7541-11eb-889a-34522b8afaa2.png)

Having a folder as the root of a prefab causes the flattening procedure to destroy all children. Maybe it should leave it for when instantiated, then flatten itself onto the hierarchy as expected?

(2019.4, 2020.1.4f1, 2021.1.0b7)

SolidAlloy replied:
If the prefab will flatten upon instantiation, it will be against the purpose of the plugin which is to improve the game performance. Flattening will slow down the game instead of speeding it up. I suggest the following alternative:
1. If the folder is a root of the prefab, a warning will be output, saying that the hierarchy will not flatten for this prefab.
2. The Folder component will be stripped from the folder game object, so just an empty game object will be left as a prefab root.

What do you think?

xsduan replied:
That sounds good to me. Let's do it that way.

SolidAlloy replied:
Added the following warning:

> Hierarchy will not flatten for {prefabPath} because its root is a folder. It's advised to make the root an empty game object.

Let me know if the phrasing is fine.

xsduan commented on Editor/Settings/SettingsDrawer.cs:
```diff
@@ -0,0 +1,111 @@
+
+            EditorGUILayout.Space(EditorGUIUtility.singleLineHeight);
+
+            EditorGUILayout.HelpBox(
+                "If you notice that entering play mode takes too long, you can try disabling this option. " +
+                "Folders will not be stripped from prefabs that are instantiated at runtime, but if performance in " +
+                "Play Mode does not matter, you will be fine.", MessageType.Warning);
```
Maybe this could only be shown when enabled, and of a message type of info

SolidAlloy replied:
You are right, it's better behaviour :) Changed in 2a3c6a184e4524b923306eb6748936fa0af15701

xsduan reviewed (changes requested):
Thank you so much for this feature!

I did find a few issues though. I can go fix them if you want, but let me know if they're easy to fix.

SolidAlloy committed:
Fixed null-reference exception when domain reload is enabled in Play Mode options

SolidAlloy committed:
Fixed Play Mode prefabs setting getting reset to whatever the build prefabs setting is

SolidAlloy committed:
Started hiding the prefab Play Mode message in Preferences and changed its type to info

SolidAlloy committed:
Fixed changed prefabs not serializing occasionally

Sometimes play mode options are disabled but DisableDomainReload flag is still set, so we need to check if the options are enabled

SolidAlloy committed:
Started grouping assetdatabase imports when stripping folders from prefabs

- Introduced the AssetImportGrouper class to make it easier to work with AssetDatabase.StartAssetEditing()
- Wrapped prefab folder-stripping methods into AssetImportGrouper
- Added AssetDatabase.Refresh at the end of RevertChanges to make the editor show changes instantly after exiting Play Mode.

SolidAlloy committed:
Fixed NullReferenceException when scripts are changed during PlayMode.

ExitingPlayMode executes before a domain reload caused by the scripts change during PlayMode, thus fixing the issue.

SolidAlloy committed:
Stopped trying to flatten the prefab hierarchy if the root game object is a folder

SolidAlloy committed:
Fixed double separators appearing in prefab instances

Flatten is called twice for prefab instances present in the scene: first when prefab is edited, and second when the scene hierarchy is processed. This caused double --- wrappers appearing when "Replace with separator" option was chosen.

SolidAlloy committed:
Removed debug logs that were accidentally left in the code

Oops

SolidAlloy committed:
Started using guids instead of paths for edited prefabs

This is done to properly handle cases when prefabs are moved or deleted in Play Mode.
